cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i

```diff
@@ -187,20 +188,23 @@ private:
         if (!isPositive<algorithmFPType, cpu>(imp.var)) imp.var = 0; //roundoff error
     }
 
-    static void calcPrevImpurity(ImpurityData & imp, ClassIndexType x, size_t n, size_t l)
+    // TODO: check for weight
+    static void calcPrevImpurity(ImpurityData & imp, ClassIndexType x, algorithmFPType nTotalWeight, algorithmFPType l)
     {
-        algorithmFPType delta = (2. * algorithmFPType(n) - algorithmFPType(l)) * imp.var + 2. * (algorithmFPType(imp.hist[x]) - algorithmFPType(n));
-        imp.var += algorithmFPType(l) * delta / (algorithmFPType(n - l) * algorithmFPType(n - l));
+        algorithmFPType delta = (2. * nTotalWeight - l) * imp.var + 2. * (algorithmFPType(imp.hist[x]) - nTotalWeight);
+        imp.var += l * delta / ((nTotalWeight - l) * (nTotalWeight - l));
         imp.hist[x] -= l;
     }
-    static void flush(ImpurityData & left, ImpurityData & right, ClassIndexType xi, size_t n, size_t k, size_t & ll)
+
+    // TODO: check for weight
+    static void flush(ImpurityData & left, ImpurityData & right, ClassIndexType xi, algorithmFPType nTotalWeight, algorithmFPType k,
+                      algorithmFPType & ll)
```

Comment:
k and ll need to have better names in this context

Reply:
It'll be done later